multi/stm32l4-multi/lib/libuart.c

```diff
@@ -0,0 +1,310 @@
+			*(ctx->base + cr1) |= 1 << 9;
+		else
+			*(ctx->base + cr1) &= ~(1 << 9);
+
+		*(ctx->base + icr) = -1;
+		(void)*(ctx->base + rdr);
+
+		if (enable) {
+			*(ctx->base + cr1) |= (1 << 5) | (1 << 3) | (1 << 2);
+			dataBarier();
+			*(ctx->base + cr1) |= 1;
+			ctx->enabled = 1;
+		}
+
+		dataBarier();
+	}
+
+	mutexUnlock(ctx->lock);
+	mutexUnlock(ctx->rxlock);
+	mutexUnlock(ctx->txlock);
+
+	return err;
+}
+
+
+int libuart_write(libuart_ctx *ctx, void* buff, unsigned int bufflen)
+{
+	if (!bufflen)
+		return 0;
+
+	if (!ctx->enabled)
+		return -EIO;
+
+	mutexLock(ctx->txlock);
+	mutexLock(ctx->lock);
+
+	keepidle(1);
+
+	*(ctx->base + tdr) = *((unsigned char *)buff);
+	ctx->txbeg = (void *)((unsigned char *)buff + 1);
+	ctx->txend = (void *)((unsigned char *)buff + bufflen);
+	*(ctx->base + cr1) |= 1 << 7;
+
+	while (ctx->txbeg != ctx->txend)
+		condWait(ctx->txcond, ctx->lock, 0);
+	mutexUnlock(ctx->lock);
+
+	keepidle(0);
+	mutexUnlock(ctx->txlock);
+
+	return bufflen;
+}
+
+
+int libuart_read(libuart_ctx *ctx, void* buff, unsigned int count, char mode, unsigned int timeout)
+{
+	int i, err;
+	volatile unsigned int read = 0;
+	char mask = 0x7f;
+
+	if (!ctx->enabled)
+		return -EIO;
+
+	if (!count)
+		return 0;
+
+	mutexLock(ctx->rxlock);
+	mutexLock(ctx->lock);
+
+	ctx->read = &read;
+	ctx->rxend = (char *)buff + count;
+
+	/* This field works as trigger for rx interrupt to store data in buffer
+	 * instead of FIFO */
+	ctx->rxbeg = buff;
+
+	/* Provoke UART exception to fire so that existing data from
+	 * rxdfifo is copied into buff. The handler will clear this
+	 * bit. */
+
+	*(ctx->base + cr1) |= 1 << 7;
+
+	while (ctx->rxbeg != ctx->rxend) {
+		err = condWait(ctx->rxcond, ctx->lock, timeout);
+
+		if (mode == uart_mnblock || (timeout && err == -ETIME) || !ctx->enabled) {
+			ctx->rxbeg = NULL;
+			ctx->rxend = NULL;
+			ctx->read = NULL;
+			break;
+		}
+	}
+
+	if (ctx->bits < 8) {
+		if (ctx->bits == 6)
+			mask = 0x3f;
+
+		for (i = 0; i < read; ++i)
+			((char *)buff)[i] &= mask;
+	}
+	mutexUnlock(ctx->lock);
+
+	mutexUnlock(ctx->rxlock);
+
+	return read;
+}
+
+
+int libuart_init(libuart_ctx *ctx, int uart)
+{
+
+	const struct {
+		volatile uint32_t *base;
+		int dev;
+		unsigned irq;
+	} info[] = {
+		{ (void *)0x40013800, pctl_usart1, usart1_irq },
+		{ (void *)0x40004400, pctl_usart2, usart2_irq },
+		{ (void *)0x40004800, pctl_usart3, usart3_irq },
+		{ (void *)0x40004c00, pctl_uart4, uart4_irq },
+		{ (void *)0x40005000, pctl_uart5, uart5_irq },
+		{ (void *)0x40008000, pctl_lpuart1, lpuart1_irq }
+	};
+
+	if (uart < 0 || uart > (int)(sizeof(info) / sizeof(info[0])))
+		return -1;
+
+	devClk(info[uart].dev, 1);
+
+	mutexCreate(&ctx->rxlock);
+	condCreate(&ctx->rxcond);
+	mutexCreate(&ctx->txlock);
+	condCreate(&ctx->txcond);
+
+	mutexCreate(&ctx->lock);
+
+	ctx->base = info[uart].base;
+
+	ctx->txbeg = NULL;
+	ctx->txend = NULL;
+
+	ctx->rxbeg = NULL;
+	ctx->rxend = NULL;
+	ctx->read = NULL;
+	ctx->rxdr = 0;
+	ctx->rxdw = 0;
```

Comment:
Initalization is in `libuart_configure()`

Reply:
Fixed, thanks